packages/loopring_v3/contracts/amm/libamm/AmmBlockReceiver.sol

```diff
@@ -27,67 +29,97 @@ library AmmBlockReceiver
             exchangeDepositContract: address(S.exchange.getDepositContract()),
             txIdx: txIdx,
             domainSeparator: S.domainSeparator,
             exchangeDomainSeparator: S.exchange.getDomainSeparator(),
+            accountID: S.accountID,
             ammActualL2Balances: new uint96[](size),
             ammExpectedL2Balances: new uint96[](size),
             numTransactionsConsumed: 0,
             tokens: S.tokens,
             size: size,
             poolTokenBase: AmmData.LP_TOKEN_BASE(),
-            poolTokenInitialSupply: AmmData.LP_TOKEN_INITIAL_SUPPLY()
+            poolTokenInitialSupply: AmmData.LP_TOKEN_INITIAL_SUPPLY(),
+            poolAmountToBurn: 0
         });
 
         BlockReader.BlockHeader memory header = _block.readHeader();
         require(header.exchange == address(ctx.exchange), "INVALID_EXCHANGE");
 
-        // The openning AMM updates
-        // This also pulls the AMM balances onchain.
+        // The openning AMM updates. This also pulls the AMM balances onchain.
         S.processAmmUpdates(ctx, true);
 
-        // Process all pool transactions
-        for (uint i = 0; i < poolTransactions.length; i++) {
-            _processPoolTransaction(S, ctx, poolTransactions[i]);
+        // Process all pool transactions and get a list of pool token transfers from
+        // the pool to newly-joined users.
+        uint txCount = poolTransactions.length;
+        AmmData.PoolTokenTransfer[] memory poolTokenTransfers = new AmmData.PoolTokenTransfer[](txCount);
+
+        for (uint i = 0; i < txCount; i++) {
+            poolTokenTransfers[i] = _processPoolTransaction(S, ctx, poolTransactions[i]);
         }
 
-        // Deposit/Withdraw to/from the AMM account when necessary
+        // Deposit to or withdraw from the AMM account when necessary, this includes the pool token.
         for (uint i = 0; i < size; i++) {
             _processPoolBalance(
-                S,
                 ctx,
                 ctx.tokens[i],
                 ctx.ammExpectedL2Balances[i],
                 ctx.ammActualL2Balances[i]
             );
         }
 
+        // Now that pool tokens may have been deposited to the pool's layer-2 account,
+        // we need to transfer these pool tokens to the user's address.
+        for (uint i = 0; i < txCount; i++) {
+            AmmJoinProcess.processPoolTokenTransfer(ctx, poolTokenTransfers[i]);
+        }
+
         // The closing AMM updates
         S.processAmmUpdates(ctx, false);
 
+        // Save the pool tokens to burn
+        S.poolAmountToBurn = ctx.poolAmountToBurn;
+
         return ctx.numTransactionsConsumed;
     }
 
+    function afterBlockSubmission(
+        AmmData.State      storage S,
+        ExchangeData.Block memory  /* _block */
+        )
+        public
+    {
+        uint poolAmountToBurn = S.poolAmountToBurn;
+        if (poolAmountToBurn > 0) {
+            S.poolAmountToBurn = 0;
+            S.burn(address(this), poolAmountToBurn);
```

Comment:
LPToken distribution may fail in Exchange, then maybe there will not be enough token to burn. So we should do the LPToken withdrawal from Exchange first.